# Binh - Chapter 3 Homework
Problems 3.1.6c, 3.1.8c, 3.2.2c, 3.3.4b, 3.3.8a, 3.4.2d, 3.4.4d: due Mon., April 13

In [15]:
import numpy as np
from matplotlib import pyplot as plt

## 3.1.6c
This exercise will be done on paper (whose pdf will be uploaded along with this ipynb).


## 3.1.8c
Similar to the previous exercise, most of this question will be done on paper. However, the numerical estimates for the values $M_1$ and $M_2$ will be computed here using numpy.

In [16]:
def f2(x):
    return 2*np.cos(x) - 4*x*np.sin(x) - x**2*np.cos(x)

def f3(x):
    return x**2*np.sin(x) - 6*x*np.cos(x) - 6*np.sin(x)

# estimating M1, max val of |f''(x)| in the 1st interv
x_vals_m1 = np.linspace(0.1, 0.2, 200000)
M1 = np.max(np.abs(f2(x_vals_m1)))

# estimate M2, max val of |f'''(x)| in the wider interv (with more draws since bigger interval)
x_vals_m2 = np.linspace(0.1, 0.3, 400000)
M2 = np.max(np.abs(f3(x_vals_m2)))

print("M1 ≈", M1)
print("M2 ≈", M2)


M1 ≈ 1.9401249222445403
M2 ≈ 3.466130101794607


## 3.2.2c

We will use the first implementation seen in the provided notebooks for this question.

In [17]:
def q(xs,ys,i,j,t):
    if j == 0:
        return ys[i]
    elif j > i:
        return 0
    else:
        return ((t-xs[i-j])*q(xs,ys,i,j-1,t) - (t-xs[i])*q(xs,ys,i-1,j-1,t))/(xs[i]-xs[i-j])

# Note that this is calculating with the entire array
def approx(xs,ys,t):
    return q(xs,ys,len(xs)-1,len(ys)-1,t)

xs = [0.1, 0.2, 0.3, 0.4]
ys = [-0.29004986, -0.56079734, -0.81401972, -1.0526302]
t = 0.18

# So when we use approx(), we actually need to filter the arrays out ourselves
p1 = approx(xs[:2], ys[:2], t)
p2 = approx(xs[:3], ys[:3], t)
p3 = approx(xs[:4], ys[:4], t)

print(p1, p2, p3)

-0.506647844 -0.508049852 -0.5081430744


The results are indistinguishable from our calculations in 3.1.6c.

## 3.3.4b

The exercise requires us to (1) find the coefficients of the relevant polynomials, and (2) evaluate $P_n(0.18)$. 
Using the formula provided, we have:
$$
P_n(x) = f(x_0) + \sum_{k=1}^n \binom{s}{k}\Delta^k f(x_0),
\qquad s = \frac{x - x_0}{h}
$$

In [18]:
# define a function that calculates the iterated differences
def forward_diffs(y):
    diffs = [np.array(y, dtype=float)]  # make an array/table
    while len(diffs[-1]) > 1:           # runs until we get delta^(number of points-1)
        diffs.append(np.diff(diffs[-1]))# finds the difference between current/previous val, appends it to array
    return diffs

# define another f that implements the forward difference formula
def newton_est(x, x0, h, diffs, degree):
    s = (x - x0) / h
    p = diffs[0][0]                     # p is initially just f(x0)
    term = 1.0                          # since s choose 0 is 1
    for k in range(1, degree + 1):
        term *= (s - (k - 1)) / k       # this is supposed to mimic (s choose k) for each k value
        p += term * diffs[k][0]         # each loop adds (s choose k) * delta^k f(x0)
    return p

x = [0.1, 0.2, 0.3, 0.4]
y = [-0.29004986, -0.56079734, -0.81401972, -1.0526302]

# first degree estimation
p41 = newton_est(0.18, x[0], x[1] - x[0], diffs, 1)
print("p1(0.18) =", p41)
print(diffs[0][0], diffs[1][0])

p1(0.18) = -0.506647844
-0.29004986 -0.27074748000000004


All claims below operate with the assumption that $s = \frac{x-0.1}{0.1}$.

**Case 1: Linear**
The output above shows the value of $P_1(0.18)$ and $f(x_0)$ and $\Delta f(x_0)$. This means that the Newton polynomial takes the form:
$$
P_1 = -0.29004986 + (-0.27074748) \times s
$$

In [19]:
# Quadratic
p42 = newton_est(0.18, x[0], x[1] - x[0], diffs, 2)
print("p2(0.18) =", p42)
print(diffs[0][0], diffs[1][0], diffs[2][0])

p2(0.18) = -0.508049852
-0.29004986 -0.27074748000000004 0.01752510000000007


**Case 2: Quadratic**
The output above shows the value of $P_2(0.18)$, $f(x_0)$, $\Delta f(x_0)$, and $\Delta^2 f(x_0)$. This means that the Newton polynomial takes the form:
$$
P_2 = -0.29004986 + (-0.27074748) \times s + (0.01752510) \times \frac{s(s-1)}{2}
$$

In [20]:
# Cubic
p43 = newton_est(0.18, x[0], x[1] - x[0], diffs, 3)
print("p3(0.18) =", p43)
print(diffs[0][0], diffs[1][0], diffs[2][0], diffs[3][0])

p3(0.18) = -0.5081430744
-0.29004986 -0.27074748000000004 0.01752510000000007 -0.002913200000000171


**Case 3: Cubic**
The output above shows the value of $P_3(0.18)$, $f(x_0)$, $\Delta f(x_0)$, $\Delta^2 f(x_0)$, and $\Delta^3 f(x_0)$. This means that the Newton polynomial takes the form:
$$
P_3 = -0.29004986 + (-0.27074748) \times s + (0.01752510) \times \frac{s(s-1)}{2} + (-0.00291320) \times \frac{s(s-1)(s-2)}{6}
$$

It is important to note that the results of $P_n(0.18)$ matches exactly with our estimates from 3.1.6c

## 3.3.8a

We use the functions provided to calculate the coefficients. Since the table has 5 pairs of x-y values, a degree-4 polynomial is needed to interpolate them.

In [21]:
# Declaring newton polynomial functions.
def newton_coeffs(xs,ys):
    n = len(xs)
    t = np.zeros((n,n))
    for i,y in enumerate(ys):
        t[i][0] = y
    for col in range(1,n):
        for row in range(col,n):
            t[row][col] = (t[row][col-1]-t[row-1][col-1])/(xs[row]-xs[row-col])
    return [t[i][i] for i in range(0,n)]
    
xs = np.array([0.0, 0.1, 0.3, 0.6, 1.0])
ys = np.array([-6.00000, -5.89483, -5.65014, -5.17788, -4.28172])
a4 = newton_coeffs(xs,ys)
print(a4)

[np.float64(-6.0), np.float64(1.051700000000002), np.float64(0.5724999999999832), np.float64(0.21500000000004166), np.float64(0.06301587301582073)]


With the coefficients calculated, we can write the polynomial out:
$$
P_4(x)= -6 + 1.0517x + 0.5725x(x-0.1) + 0.215x(x-0.1)(x-0.3) \\
+ 0.0630158730158x(x-0.1)(x-0.3)(x-0.6)
$$
with $x-0$ collapsed into $x$.

## 3.4.2d

Using the provided functions, we implement the Hermite polynomial interpolation process below:

In [22]:
def hermite_coeffs(xs,ys,yps):
    n = len(xs)
    t = np.zeros((2*n,2*n))
    for i in range(0,n):
        t[2*i][0] = ys[i]
        t[2*i+1][0] = ys[i]
    for i in range(1,n):
        t[2*i][1] = (t[2*i][0]-t[2*i-1][0])/(xs[i]-xs[i-1])
        t[2*i-1][1] = yps[i-1]
    t[2*n-1][1] = yps[n-1]
    for col in range(2,2*n):
        for row in range(col,2*n):
            t[row][col] = (t[row][col-1]-t[row-1][col-1])/(xs[row//2]-xs[(row-col)//2])
    return [t[i][i] for i in range(0,2*n)]

xs = np.array([-1, -0.5, 0, 0.5])
ys = np.array([0.86199480, 0.95802009, 1.0986123, 1.2943767])
yps = np.array([0.15536240, 0.23269654, 0.33333333, 0.45186776])

a5 = hermite_coeffs(xs,ys,yps)
print(a5)
print(len(a5))

[np.float64(0.8619948), np.float64(0.1553624), np.float64(0.07337636000000014), np.float64(0.01583111999999942), np.float64(-0.00014727999999886165), np.float64(-0.000892440000003436), np.float64(-7.6719999995984e-05), np.float64(6.863999999666782e-05)]
8


From the text (specifically algorithm 3.3), we know that the function returns an array that allows us to construct the polynomial in a specific form:
$$
\begin{aligned}
H(x) ={}& 0.8619948 + 0.1553624(x-x_0) + 0.07337636000000014(x-x_0)^2 \\
&+ 0.01583111999999942(x-x_0)^2(x-x_1) \\
&- 0.00014727999999886165(x-x_0)^2(x-x_1)^2 \\
&- 0.000892440000003436(x-x_0)^2(x-x_1)^2(x-x_2) \\
&- 0.000076719999995984(x-x_0)^2(x-x_1)^2(x-x_2)^2 \\
&+ 0.00006863999999666782(x-x_0)^2(x-x_1)^2(x-x_2)^2(x-x_3)
\end{aligned}
$$

Substituting relevant values of $x$ back in, and truncating some of the coefficients and float artifacts leaves us with:
$$
\begin{aligned}
H(x) ={}& 0.8619948 + 0.1553624(x+1) + 0.07337636(x+1)^2 \\
&+ 0.01583112(x+1)^2(x+0.5) \\
&- 0.00014728(x+1)^2(x+0.5)^2 \\
&- 0.00089244(x+1)^2(x+0.5)^2x \\
&- 0.00007672(x+1)^2(x+0.5)^2x^2 \\
&+ 0.00006864(x+1)^2(x+0.5)^2x^2(x-0.5)
\end{aligned}
$$

## 3.4.4d

Thankfully, we do not need to write the Hermite polynomial's coefficients out every time to calculate our expected values. Using the function below, we can calculate $H(0.25)$, and find the absolute error between the data generating function $f(x) = \ln(e^x + 2)$: 

In [23]:
def make_hermite_poly(xs,ys,yps):
    c = hermite_coeffs(xs,ys,yps)
    n = len(c)
    def f(t):
        sum = 0
        for i in range(0,n):
            prod = c[i]
            for j in range(0,i):
                prod = prod*(t-xs[j//2])
            sum = sum + prod
        return sum
    return f

ph = make_hermite_poly(xs,ys,yps)

# defining the target function
def f(x):
    return np.log(np.exp(x)+2)

print("Absolute error is approximately", np.abs(f(0.25)-ph(0.25)))

Absolute error is approximately 1.699302785151957e-07


As we can see, the Hermite polynomial seems to track the target function quite effectively.